### ISSUE TITLE
[G2] PSA crypto should not depend on the cipher module

### ISSUE THREAD

**Issue by gilles-peskine-arm** [Mbed-TLS/mbedtls#8048]
As of Mbed TLS 3.4, `MBEDTLS_PSA_CRYPTO_C` is registered as depending on `MBEDTLS_CIPHER_C` in `check_config.h`. This is obsolete, dating back from before we had driver support for ciphers. Now, if there are no cipher/AEAD algorithms that use the built-in implementation, PSA doesn't need `MBEDTLS_CIPHER_C`.

At the time of writing, the build-in implementation of all non-authenticated ciphers depends heavily on the cipher module. On the other hand, the build-in implementation of AEAD only directly depends on cipher for metadata conversion. Then the built-in GCM and CCM modules themselves depend on cipher, but chachapoly doesn't. So it should be easy to make a cipher-less chachapoly-only build possible.

The goal of this task is:

* In `psa_aead_setup`, only call `mbedtls_cipher_info_from_psa` if `MBEDTLS_CIPHER_C` is defined.
* Change `check_config.h` to only require `MBEDTLS_CIPHER_C` based on `MBEDTLS_PSA_CRYPTO_C` if the built-in implementation of one of (a non-authenticated cipher, CCM, GCM) is enabled (`PSA_HAVE_SOFT_CIPHER || MBEDTLS_PSA_BUILTIN_ALG_CCM || MBEDTLS_PSA_BUILTIN_ALG_GCM`, where `PSA_HAVE_SOFT_CIPHER` needs to be defined in `config_psa.h` or its successor).
* Make sure that when `MBEDTLS_PSA_CRYPTO_CONFIG` is enabled, `config_psa.h` or its successor enables `MBEDTLS_CIPHER_C` when required.
* In `component_test_full_no_cipher`, change `test_full_no_cipher` to keep PSA enabled.

Follow-up:

* https://github.com/Mbed-TLS/mbedtls/issues/6317: add or change a build with all ciphers accelerated to have `MBEDTLS_CIPHER_C` disabled.

**Comment by valeriosetti** [comment]
I started to work on this. I have some doubt about this part:
> but chachapoly doesn't. So it should be easy to make a cipher-less chachapoly-only build possible.

If we guard `mbedtls_cipher_info_from_psa` in `psa_aead_setup` with `MBEDTLS_CIPHER_C` then:
- `test_suite_psa_crypto` passes all tests
- `test_suite_psa_crypto_op_fail.generated` does not

The reason is that in following tests:
```
PSA aead CHACHA20_POLY1305: incompatible with DERIVE .............. FAILED
PSA aead CHACHA20_POLY1305: incompatible with DH_KEY_PAIR(RFC7919)  FAILED
PSA aead CHACHA20_POLY1305: incompatible with DH_PUBLIC_KEY(RFC791  FAILED
PSA aead CHACHA20_POLY1305: incompatible with HMAC ................ FAILED
PSA aead CHACHA20_POLY1305: incompatible with PASSWORD ............ FAILED
PSA aead CHACHA20_POLY1305: incompatible with PASSWORD_HASH ....... FAILED
PSA aead CHACHA20_POLY1305: incompatible with RAW_DATA ............ FAILED
PSA aead CHACHA20_POLY1305: incompatible with RSA_KEY_PAIR ........ FAILED
PSA aead CHACHA20_POLY1305: incompatible with RSA_PUBLIC_KEY ...... FAILED
```
we are trying to set an unsupported key type which is not rejected by `mbedtls_cipher_info_from_psa`.

If I do the opposite and I only guard `mbedtls_cipher_info_from_values` in `mbedtls_cipher_info_from_psa` with `MBEDTLS_CIPHER_C`, then `test_suite_psa_crypto_op_fail.generated` passes tests while `test_suite_psa_crypto` does not. The reason in this case is that the program doesn't reach the proper `case` in `psa_aead_setup` because the cipher is rejected in `mbedtls_cipher_info_from_psa`.

In conclusion, albeit it is true that is possible to build with chacha supported in PSA, then this support is incomplete. I think we need something like `CIPHER_LIGHT` to have partial support for ciphers, like understanding the algorithm and key, without math. However we don't have `CIPHER_LIGHT` for now and it is the target of [another task](https://github.com/Mbed-TLS/mbedtls/issues/8058) in the future.

My proposal here is to build removing chacha support as well. @gilles-peskine-arm wdyt?

**Comment by gilles-peskine-arm** [comment]
@valeriosetti I don't understand exactly what you tried. I expect to be able to enable `MBEDTLS_PSA_CRYPTO_C` and `PSA_WANT_KEY_TYPE_CHACHA20` and `PSA_WANT_ALG_CHACHA20_POLY1305` without having `MBEDTLS_CIPHER_C` enabled (and using the built-in implementation for all algorithms). Note that I don't expect this to be possible if `PSA_WANT_ALG_CHACHA20` is enabled because `psa_crypto_cipher.c` depends on `cipher.c` a lot. Since `MBEDTLS_CHACHAPOLY_1305_C` requires `MBEDTLS_CHACHA20_C`, the only way to reach this configuration is with `MBEDTLS_PSA_CRYPTO_CONFIG` enabled, so that we can have `MBEDTLS_CHACHA20_C` without `PSA_WANT_ALG_CHACHA20`.

https://github.com/Mbed-TLS/mbedtls/pull/8119 looks like a good first step, with a second step of having a test that enables chacha20-poly1305. I'm not sure if `component_test_full_no_cipher` should switch to using `MBEDTLS_PSA_CRYPTO_CONFIG` or if we should have a separate component for that.

**Comment by valeriosetti** [comment]
> @valeriosetti I don't understand exactly what you tried. I expect to be able to enable MBEDTLS_PSA_CRYPTO_C and PSA_WANT_KEY_TYPE_CHACHA20 and PSA_WANT_ALG_CHACHA20_POLY1305 without having MBEDTLS_CIPHER_C enabled (and using the built-in implementation for all algorithms).

The first 7 commits (up to https://github.com/Mbed-TLS/mbedtls/pull/8119/commits/055f21656fefbb00d9c079e1bfa23e60145c225f) of #8119 implement what you described: it get `test_full_no_cipher` to a point where it has support for ChaCha in PSA without `CIPHER_C`. However, as stated in [previous comment](https://github.com/Mbed-TLS/mbedtls/issues/8048#issuecomment-1692826911), I have problems with `test_suite_psa_crypto_op_fail.generated` test suite.
I did not enable `MBEDTLS_PSA_CRYPTO_CONFIG` in that test component, so I can try to enable it and see if that somehow helps in fixing the problem.

**Comment by mpg** [comment]
I'm wondering it we might resolve this by splitting `mbedtls_cipher_info_from_psa()` into:
1. A new function, say `mbedtls_cipher_values_from_psa()` that does not depend on cipher at all, and is basically the same as the current `mbedtls_cipher_info_from_psa()` except it does not call `mbedtls_cipher_info_from_values()` at the end but returns `cipher_id`, `key_bits` and `mode` instead.
2. The call to `mbedtls_cipher_info_from_values()` which is the only part the depends on `CIPHER_C` (and soon only `CIPHER_LIGHT` indeed, but that's quite orthogonal: ideally, support for ChachaPoly in PSA should not need even `CIPHER_LIGHT`).

Then the new function from (1) could be used unconditionally in `psa_aead_setup()` to ensure that the key type and attributes are correct, and then we'd call `mbedtls_cipher_info_from_values()` only on the GCM and CCM code paths, which would be the only ones depending on `CIPHER_C` (soon `CIPHER_LIGHT`).

Wdyt? Am I missing something?

**Comment by mpg** [comment]
> I'm not sure if `component_test_full_no_cipher` should switch to using `MBEDTLS_PSA_CRYPTO_CONFIG` or if we should have a separate component for that.

I have a strong preference for leaving `test_full_no_cipher` as it is, and adding another one with `MBEDTLS_PSA_CRYPTO_CONFIG` enabled, as well as `PSA_CRYPTO_C`, but disabling everything we need to disable in order to make `PSA_CRYPTO_C` no re-enable `CIPHER_C` under the hood.

**Comment by gilles-peskine-arm** [comment]
> I'm wondering it we might resolve this by splitting mbedtls_cipher_info_from_psa() into:

> A new function, say mbedtls_cipher_values_from_psa() that does not depend on cipher at all, and is basically the same as the current mbedtls_cipher_info_from_psa() except it does not call mbedtls_cipher_info_from_values() at the end but returns cipher_id, key_bits and mode instead.

For AEAD, all that's needed is the `cipher_id`, to pass to the gcm and ccm modules. No code will use the mode, and PSA knows the key size independently.

> The call to mbedtls_cipher_info_from_values() which is the only part the depends on CIPHER_C (and soon only CIPHER_LIGHT indeed, but that's quite orthogonal: ideally, support for ChachaPoly in PSA should not need even CIPHER_LIGHT).

Calling cipher or cmac requires an info structure. It would be more complex to re-assemble that from the separate legacy values than to directly obtain it from the PSA values.
